File: packages/react-native-vision-camera-barcode-scanner/ios/Extensions/ML+BarcodeFormat.swift
//
//  ML+TargetBarcodeFormat.swift
//  VisionCameraBarcodeScanner
//
//  Created by Marc Rousavy on 08.02.26.
//

import MLKitBarcodeScanning

extension BarcodeFormat {
  init(fromMLKitFormat format: MLKitBarcodeScanning.BarcodeFormat) {
    switch format {
    case .code128:
      self = .code128
    case .code39:
      self = .code39
    case .code93:
      self = .code93
    case .codaBar:
      self = .codabar
    case .dataMatrix:
      self = .dataMatrix
    case .EAN13:
      self = .ean13
    case .EAN8:
      self = .ean8
    case .ITF:
      self = .itf
    case .qrCode:
      self = .qrCode
    case .UPCA:
      self = .upcA
    case .UPCE:
      self = .upcE
    case .PDF417:
      self = .pdf417
    case .aztec:
      self = .aztec
    default:
      self = .unknown
    }
  }
}

extension TargetBarcodeFormat {
  func toMLKitFormat() -> MLKitBarcodeScanning.BarcodeFormat {
    switch self {
    case .allFormats:
      return .all
    case .code128:
      return .code128
    case .code39:
      return .code39
    case .code93:
      return .code93
    case .codabar:
      return .codaBar
    case .dataMatrix:
      return .dataMatrix
    case .ean13:
      return .EAN13
    case .ean8:
      return .EAN8
    case .itf:
      return .ITF
    case .qrCode:
      return .qrCode
    case .upcA:
      return .UPCA
    case .upcE:
      return .UPCE
    case .pdf417:
      return .PDF417
    case .aztec:
      return .aztec
    }
  }
}

File: packages/react-native-vision-camera-barcode-scanner/android/src/main/java/com/margelo/nitro/camera/barcodescanner/extensions/ML+BarcodeFormat.kt
package com.margelo.nitro.camera.barcodescanner.extensions

import com.google.mlkit.vision.barcode.common.Barcode
import com.margelo.nitro.camera.barcodescanner.BarcodeFormat
import com.margelo.nitro.camera.barcodescanner.TargetBarcodeFormat

@Barcode.BarcodeFormat
fun TargetBarcodeFormat.toMLBarcodeFormat(): Int {
  return when (this) {
    TargetBarcodeFormat.CODE_128 -> Barcode.FORMAT_CODE_128
    TargetBarcodeFormat.CODE_39 -> Barcode.FORMAT_CODE_39
    TargetBarcodeFormat.CODE_93 -> Barcode.FORMAT_CODE_93
    TargetBarcodeFormat.CODABAR -> Barcode.FORMAT_CODABAR
    TargetBarcodeFormat.DATA_MATRIX -> Barcode.FORMAT_DATA_MATRIX
    TargetBarcodeFormat.EAN_13 -> Barcode.FORMAT_EAN_13
    TargetBarcodeFormat.EAN_8 -> Barcode.FORMAT_EAN_8
    TargetBarcodeFormat.ITF -> Barcode.FORMAT_ITF
    TargetBarcodeFormat.QR_CODE -> Barcode.FORMAT_QR_CODE
    TargetBarcodeFormat.UPC_A -> Barcode.FORMAT_UPC_A
    TargetBarcodeFormat.UPC_E -> Barcode.FORMAT_UPC_E
    TargetBarcodeFormat.PDF_417 -> Barcode.FORMAT_PDF417
    TargetBarcodeFormat.AZTEC -> Barcode.FORMAT_AZTEC
    TargetBarcodeFormat.ALL_FORMATS -> Barcode.FORMAT_ALL_FORMATS
  }
}

fun BarcodeFormat.Companion.fromMLBarcodeFormat(
  @Barcode.BarcodeFormat format: Int,
): BarcodeFormat {
  return when (format) {
    Barcode.FORMAT_UNKNOWN -> BarcodeFormat.UNKNOWN
    Barcode.FORMAT_CODE_128 -> BarcodeFormat.CODE_128
    Barcode.FORMAT_CODE_39 -> BarcodeFormat.CODE_39
    Barcode.FORMAT_CODE_93 -> BarcodeFormat.CODE_93
    Barcode.FORMAT_CODABAR -> BarcodeFormat.CODABAR
    Barcode.FORMAT_DATA_MATRIX -> BarcodeFormat.DATA_MATRIX
    Barcode.FORMAT_EAN_13 -> BarcodeFormat.EAN_13
    Barcode.FORMAT_EAN_8 -> BarcodeFormat.EAN_8
    Barcode.FORMAT_ITF -> BarcodeFormat.ITF
    Barcode.FORMAT_QR_CODE -> BarcodeFormat.QR_CODE
    Barcode.FORMAT_UPC_A -> BarcodeFormat.UPC_A
    Barcode.FORMAT_UPC_E -> BarcodeFormat.UPC_E
    Barcode.FORMAT_PDF417 -> BarcodeFormat.PDF_417
    Barcode.FORMAT_AZTEC -> BarcodeFormat.AZTEC
    else -> BarcodeFormat.UNKNOWN
  }
}
